ComponentKit/Utilities/CKComponentAction.h

```diff
@@ -30,25 +23,184 @@ typedef NS_ENUM(NSUInteger, CKComponentActionSendBehavior) {
 };
 
 /**
- Sends a component action up the responder chain by crawling up the responder chain until it finds a responder that
- responds to the action's selector, then invokes it.
+ CKTypedComponentAction is a struct that represents a method invocation that can be passed to a child component to
+ trigger a method invocation on a target.
+ 
+ We allow a typed specification of the parameters that will be provided as arguments to the component action at runtime
+ through the variadic templated arguments. You may specify an arbitrary number of arguments to your component action,
+ and you may use either object, or primitive arguments. You should not use C++ references as arguments because the block
+ API will likely result in corrupted memory access. Pass C++ structs by value as arguments to action sending.
+
+ Methods will always be provided the sender as the first argument.
+ 
+ Usage in your component header:
+ 
+ @interface MyComponent : CKComponent
+ + (instancetype)newWithAction:(const CKTypedComponentAction<NSString *, int> &)action;
+ @end
+ 
+ When creating the action:
+ 
+ Option 1 - Promoted raw-selector component action. Uses a raw selector which traverses upwards looking for a parent
+            that implements methodWithNoArguments, and calls that method without any arguments. We support actions which
+            accept fewer arguments than defined in the declaration of the action above. However, types of received
+            parameters should be the same as the declaration, if they're present.
+
+ [MyComponent newWithAction:{@selector(methodWithNoArguments)}];
+ ...
+ - (void)methodWithNoArguments {}
+
+ 
+ Option 2 - Exact raw-selector component action. Uses a raw selector which also traverses upwards using the responder
+            chain, but accepts the arguments defined in the interface.
```

Comment:
aren't option 1 and option 2 the same? the rule is "you can supply up to exactly as many parameters as the action takes, or any small number of arguments." or is it truly all-or-nothing—you must take all arguments or no arguments?